docs/man/man1/xyzab-tdr-gui.1
.\" Copyright (c) 2023 andypugh
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
.\" USA.
.\"
.\"
.\"
.TH xyzab-tdr-gui "1"  "2023-07-13" "LinuxCNC Documentation" "The Enhanced Machine Controller"
.SH NAME
xyzab-tdr-gui \- Vismach Virtual Machine GUI

.SH DESCRIPTION
\fBxyzab-tdr-gui\fR is one of the sample \fBVismach\fR GUIs for LinuxCNC,
simulating a dual-rotary machine with switchable kinematics.

See the main LinuxCNC documentation for more details.

http://linuxcnc.org/docs/html/gui/vismach.html
.
.SH "SEE ALSO"
\fBLinuxCNC(1)\fR

Much more information about LinuxCNC and HAL is available in the LinuxCNC
and HAL User Manuals, found at /usr/share/doc/LinuxCNC/.

.SH BUGS
None known at this time.
.PP
.SH AUTHOR
This man page written by andypugh, as part of the LinuxCNC project.
.SH REPORTING BUGS
Report bugs at https://github.com/LinuxCNC/linuxcnc/issues
.SH COPYRIGHT
Copyright \(co 2023 andypugh.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


docs/NEWS
check debian/changelog for recent news

# some of EMC2 History & News
#############################

2009.06.13 - mshaver added emc2/configs/smithy directory and related files
             for Smithy Company's line of cnc machines.

2006.11.17 - removed HAL types s8, u8, s16, and u16.  HAL now uses float
	     for analog values, bit for booleans, and s32 or u32 for
	     integer values.  The only users who will be affected are
	     those with Universal Stepper Controller cards.  See item 9 at
	     http://wiki.linuxcnc.org/cgi-bin/emcinfo.pl?UpdatingConfigurationsForDevelopmentVersions

2006.11.15 - merged classicladder 0.7.100

2006.11.04 - added a feed-hold HAL pin.  Added M codes to enable/disable
	     feed override, spindle override, feed hold, and adaptive feed
	     individually.  Feed override is M50, spindle override is M51,
	     adaptive feed is M52, and feed hold is M53.  G50/51 are no
	     longer used to enable/disable adaptive feed, use M52 instead.

2006.10.24 - added support for block delete as defined by Section 2.2.2
             in the RS274NGC_3 specs.

2006.10.24 - added support for optional program stop as defined by Section 
             2.2.3 in the RS274NGC_3 specs.

2006.10.14 - manpages have been created for most rtapi and hal calls

2006.10.01 - added a userspace simulator which includes rtapi threads.
	     Use "--enable-simulator" to enable this mode, which cannot
	     be used to control real hardware

2006.09.27 - added support to optionally save joint positions between
             runs ([TRAJ] POSITION_FILE = position.txt)

2006.09.26 - added support to specify UNITS in literal in the ini. 
             (e.g. [TRAJ] LINEAR_UNITS = mm )

2006.09.09 - fixed bug #1384883, optional stop was not optional.

2006.08.09 - AXIS is now a part of EMC2, not a separate source download.

2006.08.07 - Added a spindle-speed override. This allows to override the spindle
	     speed during a program run. The MIN_SPINDLE_OVERRIDE and 
	     MAX_SPINDLE_OVERRIDE in the ini assure that speed doesn't get 
	     reduced or increased too much. Both values are in %.

2006.08.06 - Documentation can be built from source with the 'make docs' target

2006.06.11 - added realtime support for jogwheels.  Three HAL pins per axis:
	     one accepts raw counts from the wheel, one sets the distance moved
	     per count, and one enables that axis to move.

2006.05.17 - added 'adaptive feed', a HAL pin that can be used to scale the
	     feedrate in real time.  Simular to feedrate override, but can
	     only vary from 0.0 to 1.0, and reacts much faster.  Intended 
	     for use with EDM and other applications that require closed 
	     loop control of feedrate.

2006.03.18 - A new pin "count" exists for freqgen.  It is the same as the
             internal symbol "count" shown in the hal introduction.pdf,
             or the result of up OR down if there was an "OR" component.

2006.03.16 - Modified the motion control module to behave more like other
             HAL modules - it no longer adds its functions to threads
             automatically.  All the sample configs have been changed to
             explicitly add the functions, users who have custom configs
             will need to do the same.

2006.03.12 - If a copy of axis 1.3a0 or newer is detected inside src/
             it will be automatically built and installed for
             --enable-run-in-place use

2006.03.10 - minor bugfix in tkemc for jogging

2006.03.10 - added genhexkins, with a minitetra sample config.
             This proves EMC2 works just as good with hexapods as EMC1 did.

2006.03.10 - added VTI support & sample configs

2006.03.07 - lots of fixes to the TP (most were wrong in the old one as well)

2006.03.05 - new Trajectory Planner

2006.02.20 - added a manpage for the runscript.

2006.02.12 - Fixed a bug that was causing the RT code to run very slowly,
             forcing users to set BASE_PERIOD very long.

2006.02.10 - The build process has been replaced, no more recursive $(MAKE)

2006.01.28 - fixed bug #1205237, mode switching sometimes hung

2006.01.08 - M62/M63 digital outputs synchronous to motion
             M64/M65 digital outputs that get set right away

2006.01.06 - work on halconfig.tcl, it is used to display/configure HAL data
             requires the BWidget package, realtime, and HAL to run.

2006.01.04 - fix bug #1387740, a motion that goes nowhere clears feedrate

2006.01.01 - the config picker has i18n support, a German translation exists

2005.12.20 - a new config picker has been written, if you start just the
             runscript this one helps you choose a config to run

2005.12.13 - new runscript, and config organization
             work towards the first release of EMC2

2005.xx.xx - EMC2 now supports tool-changing in complex scenarios

2005.xx.xx - classicladder now a HAL component, a software PLC

2004.xx.xx - HAL (the Hardware Abstraction Layer) is used for
	     addressing hardware drivers, and allows for simple configuration


#developers: this file should hold a history of improvements, bugfixes, etc
             it is useful for keeping track of development


src/hal/user_comps/vismach/xyzab-tdr-gui.py
#!/usr/bin/env python3
#**************************************************************************
#
# Copyright 2023 David Mueller mueller_david@hotmail.com
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#**************************************************************************

#--------------------------------------------------------------------------
# Visualization model of a 5-axis milling machine
# with table rotary axes A and B
# This model includes dynamic adjustments of geometric offsets of
# rotational axes as well as the position of the intersection of the
# rotational axes A and B (rotation-point of the rotary-assembly).
#--------------------------------------------------------------------------

from vismach import *
import hal
import math
import sys

c = hal.component("xyzab-tdr-gui")
# axis-x
c.newpin("axis-x", hal.HAL_FLOAT, hal.HAL_IN)
# axis-y
c.newpin("axis-y", hal.HAL_FLOAT, hal.HAL_IN)
# axis-z
c.newpin("axis-z", hal.HAL_FLOAT, hal.HAL_IN)
# rotary_a
c.newpin("rotary-a", hal.HAL_FLOAT, hal.HAL_IN)
# rotary_b
c.newpin("rotary-b", hal.HAL_FLOAT, hal.HAL_IN)
# geometric offsets between A and B axes
c.newpin("x_offset", hal.HAL_FLOAT, hal.HAL_IN)
c.newpin("z_offset", hal.HAL_FLOAT, hal.HAL_IN)
# rot-point offsets
c.newpin("x-rot-point", hal.HAL_FLOAT, hal.HAL_IN)
c.newpin("y-rot-point", hal.HAL_FLOAT, hal.HAL_IN)
c.newpin("z-rot-point", hal.HAL_FLOAT, hal.HAL_IN)
# tool-length offset
# tool offsets
c.newpin("tool_length", hal.HAL_FLOAT, hal.HAL_IN)
c.newpin("tool_diameter", hal.HAL_FLOAT, hal.HAL_IN)
c.ready()

# These values are arbitrary and are meant to simulate an absolute
# machine zero that is offset from the rotation point of the rotary assembly.
# Machine zero as measured from rotation point of the rotary assembly
machine_zero_x =  50
machine_zero_y =  50
machine_zero_z =  100

for setting in sys.argv[1:]: exec(setting)

# give endpoint Z values and radii
# resulting cylinder is on the Z axis
class HalToolCylinder(CylinderZ):
    def __init__(self, comp, *args):
        CylinderZ.__init__(self, *args)
        self.comp = c

    def coords(self):
        r = 2 # default if hal pin not set
        if (c.tool_diameter > 0): r = c.tool_diameter/2
        return -c.tool_length, r, 0, r

# used to create visual indicators for the x-offset
class HalOffsetCylinderX(CylinderX):
    def __init__(self, comp, *args):
        CylinderZ.__init__(self, *args)
        self.comp = c

    def coords(self):
        length = 0 # default if hal pin not set
        length = c.x_offset
        return length, 0.1, 0, 0.3

# used to create visual indicators for the z-offset
class HalOffsetCylinderZ(CylinderZ):
    def __init__(self, comp, *args):
        CylinderZ.__init__(self, *args)
        self.comp = c

    def coords(self):
        length = 0 # default if hal pin not set
        length = c.z_offset
        return length, 0.1, 0, 0.3


# begin tool-side
tooltip = Capture()

# create the tool using the values for tool_diameter and tool_length in the tool table
tool = Collection([HalTranslate([tooltip], c, "tool_length", 0,0,-1),
                   HalToolCylinder(c),
                   ])
tool = Color([1,0,0,0], [tool] )

# create a visual indicator for the position of the "(x,y,z)-rot-point"
# so the user can see the effects of the "(x,y,z)_rot-point" value changes
rot_point_set = Collection([
                Color([1,0.8,0,1],[CylinderX(-5,0.2,5,0.2)]),
                Color([1,0.8,0,1],[CylinderY(-5,0.2,5,0.2)]),
                Color([1,0.8,0,1],[CylinderZ(-5,0.2,5,0.2)])
                ])
# Now move the rot_point indicator from the home position by the
# '(x,y,z)_rot-point' values
# For the kinematics to work correctly the 'rot_point_set' indicator must be
# shown in the same position as the red sphere indicating the rotation point
# of the rotary assembly created later
rot_point_set = HalTranslate([rot_point_set],c,"x-rot-point", 1,0,0)
rot_point_set = HalTranslate([rot_point_set],c,"y-rot-point", 0,1,0)
rot_point_set = HalTranslate([rot_point_set],c,"z-rot-point", 0,0,1)

spindle = Collection([
          # spindle nose and/or toolholder
          CylinderZ( 0, 10, 20, 15),
          # spindle housing
          CylinderZ( 20, 20, 100, 20),
          ])

head_y = Collection([
         tool,
         Color([0,0.5,0.5,0], [spindle]),
         # y-slide, holds spindle
         Color([0,1,0,0], [Box( -30, -30, 60, 30, 280, 135 )])
         ])
# add the hal connections for x,y,z axis movements
head_y = HalTranslate([head_y],c,"axis-x",1,0,0)
head_y = HalTranslate([head_y],c,"axis-y",0,1,0)
head_y = HalTranslate([head_y],c,"axis-z",0,0,1)
# move the spindle and y-slide to the simulated machine home position
head_y     = Translate([head_y], machine_zero_x, machine_zero_y, machine_zero_z)
# also move the rot_point indicator along with the spindle
# it's 'location from the spindle differs by (x,y,z)-rot-point values
rot_point_set  = Translate([rot_point_set], machine_zero_x, machine_zero_y, machine_zero_z)
# end tool-side

# begin work-side
work = Capture()

# create rotary A
table_a = Collection([
         work,
         CylinderX(-18, 50, 0, 50),
         # cross
         Color([1,1,1,0], [BoxCentered(0.1, 100, 2)]),
         Color([1,1,1,0], [BoxCentered(0.1, 2, 100)]),
         # lump on one side
         Color([1,1,1,0], [Box(-10, -2, -50, 2, 2, -40)])
         ])
# add the hal connection for the a axis rotation
table_a = HalRotate([table_a],c,"rotary-a",1,-1,0,0)
table_a = Color([1,0,1,0], [table_a] )

rotary_a = Collection([
          table_a,
          # rotary table base - block under the rotary A table
          Color([1,0.5,0,0], [Box(-50,-50, -50, -10, 100, 50)])
          ])

# simulate the 'x-offset' and 'z-offset' values
rotary_a = HalTranslate([rotary_a],c,"x_offset",1,0,0)
rotary_a = HalTranslate([rotary_a],c,"z_offset",0,0,1)

rotary_b = Collection([
       # arm
       Color([1,0.5,0,0], [CylinderY(80,50,100,50)]),
       Color([1,0.5,0,0], [Box(-70,80,-50,0,100,50)]),
       # white 'dowel' indicating the axis B rotational center
       Color([1,1,1,0], [CylinderY(75,0.2,80,2)]),
       Color([1,1,1,0], [CylinderY(0,0.1,80,0.2)]),
       #  ball at the end for the rotation-point of the rotary-assembly
       Color([1,0.1,0,0],[ Sphere(0,0,0,0.5)]),
       # show geometric offsets of the rotary-assembly
       # z-offset extending from the rotation-point
       Color([0,0,1,1],[HalOffsetCylinderZ(CylinderZ)]),
       # x-offset extending from the end of the z-offset indicator (should end at the face center of the rotary A)
       HalTranslate([Color([1,0,0,1],[HalOffsetCylinderX(CylinderX)])], c, "z_offset", 0,0,1),
       ])

dualrotary = Collection([
             rotary_a,
             rotary_b,
             ])
# add the hal connections for b axis rotation
dualrotary = HalRotate([dualrotary],c,"rotary-b",1,0,-1,0)
# end work-side

# begin base
base = Collection([
       # base
       Box(-120, -100, -250, 120, 160, -150),
       # column
       Box(-120, 100, -250, 120, 200, 300),
       ])
base = Color([0.5,0.5,0.5,0], [base] )
# end base

# begin model
model = Collection([
        head_y,
        dualrotary, base,
        rot_point_set
        ])
# end model

# begin hud
myhud = Hud()
myhud.show("xyzab-tdr-gui")
# end hud

main(model, tooltip, work, size=500, hud=myhud, lat=-60, lon=0)
